creator-node/src/routes/users.js

```diff
@@ -41,16 +53,76 @@ module.exports = function (app) {
       return errorResponseBadRequest('Invalid data or signature')
     }
 
-    // signature data is valid and matches a user, proceed with verifying timestamp is
-    // within time boundary
     const theirTimestamp = parseInt(data.split(':')[1])
     const ourTimestamp = Math.round((new Date()).getTime() / 1000)
 
     if (Math.abs(theirTimestamp - ourTimestamp) > 3600) {
       console.error(`Timestamp too old. User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
     }
 
-    // all checks have passed! generate a new session token for the user
+    const sessionToken = await sessionManager.createSession(user.cnodeUserUUID)
+    return successResponse({ sessionToken })
+  }))
+
+  /**
+   * Return a challeng used for validating user login. Challenge value
+   * is also set in redis cache with the key 'userLoginChallenge:<wallet>'.
+   */
+  app.get('/users/login/challenge', handleResponse(async (req, res, next) => {
+    const walletPublicKey = req.query.walletPublicKey
+    const userLoginChallengeKey = `userLoginChallenge:${walletPublicKey}`
+    const redisClient = req.app.get('redisClient')
+    const challengeBuffer = await randomBytes(CHALLENGE_VALUE_LENGTH)
+    const challenge = base64url.encode(challengeBuffer)
+
+    // Set challenge ttl to 2 minutes ('EX' option = sets expire time in seconds)
+    // https://redis.io/commands/set
+    // https://github.com/luin/ioredis/blob/master/examples/basic_operations.js#L44
+    await redisClient.set(userLoginChallengeKey, challenge, 'EX', CHALLENGE_TTL_SECONDS)
+
+    return successResponse({ walletPublicKey, challenge })
+  }))
+
+  /**
+   * Checks if challenge in request body matches up with what we have stored.
+   * If request challenge matches what we have, remove instance from redis to
+   * prevent replay attacks. Return sessionToken upon success.
+   */
+  app.post('/users/login/challenge', handleResponse(async (req, res, next) => {
+    const { signature, data, challenge } = req.body
+
+    if (!signature || !data || !challenge) {
+      return errorResponseBadRequest('Missing request body values.')
+    }
+
+    const address = utils.verifySignature(data, signature)
+    const user = await models.CNodeUser.findOne({
+      where: {
+        walletPublicKey: address
+      }
+    })
+    if (!user) {
+      return errorResponseBadRequest('Invalid data or signature')
+    }
+
+    const theirChallengeBytes = data.split(': ')[1]
+
+    req.logger.info(`Detected challenge ${challenge}`)
```

Comment:
@vicky-g sorry last thing, we shouldn't log this either. we shouldn't log anything to do with the challenge string either that they send over or we parse or retrieve from redis

Reply:
oh whoops missed this one. got it!